# Create Dataset

In [1]:
import numpy as np
import matplotlib.pyplot as plt

from sklearn.datasets import load_diabetes
from sklearn.model_selection import train_test_split

from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
from sklearn import set_config
set_config(display='diagram')

In [3]:
# we import the diabetes
X, y = load_diabetes(return_X_y=True)

In [4]:
print(f"Size of X :{X.shape} \nSize of y {y.shape}")

Size of X :(442, 10) 
Size of y (442,)


# Train Test Split

In [5]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=2)

In [6]:
#  I make mistake in during prediction so 
# This help me during the debugging time

print(f"Size of X_train :{X_train.shape} \nSize of y_train {y_train.shape} \n")
print(f"Size of X_test :{X_test.shape} \nSize of y_test {y_test.shape} \n")

Size of X_train :(353, 10) 
Size of y_train (353,) 

Size of X_test :(89, 10) 
Size of y_test (89,) 



# Apply Sklearn Model

In [7]:
reg = LinearRegression()

In [8]:
reg.fit(X_train, y_train)

LinearRegression()

In [9]:
reg.coef_

array([  -9.15865318, -205.45432163,  516.69374454,  340.61999905,
       -895.5520019 ,  561.22067904,  153.89310954,  126.73139688,
        861.12700152,   52.42112238])

In [10]:
reg.intercept_

np.float64(151.88331005254167)

In [11]:
y_pred = reg.predict(X_test)
y_pred.shape

(89,)

In [12]:
y_pred

array([154.1213881 , 204.81835118, 124.93755353, 106.08950893,
       258.5348576 , 256.3310074 , 118.75087616, 119.52440696,
       101.50816735, 190.54048661, 141.70656811, 172.51883961,
       174.33861649, 134.80942706, 294.13994537,  94.11798038,
       211.97059795, 156.49579378, 134.21000428, 119.62664644,
       148.87842251, 165.00873409, 151.10021038, 176.04063756,
       133.27769647, 221.29555392, 197.17324941,  96.1577688 ,
        50.26012711, 230.48580317, 242.06073866, 114.11129218,
        67.07532417,  94.52943825, 201.21415375, 167.05136201,
       159.881268  , 192.78746659, 114.49551325, 233.48234551,
       140.82563045, 121.0680409 , 192.27480772, 191.12738845,
       179.16865788, 148.34935601, 163.47414622, 276.81647884,
       100.17926432, 164.10555298, 255.80762189, 136.9466204 ,
       152.37503699, 107.92237882, 194.21924678,  77.34670792,
       118.50482479,  68.38335763, 154.29258529, 162.48840259,
       168.36788326, 156.87790322,  97.14191797, 238.16

In [13]:
r2_score(y_test, y_pred)

0.4399338661568968

# Introduce the batch size

*Batch Size* is how many apple you want from the bucket.

or 

*Batch Size* is how many rows you want from the dataset.

- here batch size is 10 
- random.sample(range(X_train.shape[0]) , 10)
- range(X_train.shape[0]) -> from 1 to 353 number 
- This sample give us 10 random number 
- why 10 
- this is our batch size



In [31]:
X_train.shape[0]

353

In [33]:
import random
random.sample(range(X_train.shape[0]) , 10)

[86, 7, 85, 80, 68, 247, 333, 299, 113, 50]

# Making own class

In [53]:
import numpy as np

class MiniBatchGDRegressor:
    
    def __init__(self, learning_rate=0.01, epochs=100, batch_size=10):
        self.coef_ = None
        self.intercept_ = None
        self.lr = learning_rate
        self.epochs = epochs
        self.batch_size = batch_size
        
    def fit(self, X, y):
        # Initialize the coefficients and intercept
        self.intercept_ = 0
        self.coef_ = np.ones(X.shape[1])
        
        # Gradient Descent Loop
        for epoch in range(self.epochs):
            
            # Select the some rows given by the user
            num = X.shape[0] / self.batch_size
            for select_rows in range(int(num)):
                # row idx randomly pick from dataset with batch size
                idx = random.sample(range(X.shape[0]), self.batch_size)
                 
                # Make predictions
                y_hat = np.dot(X[idx], self.coef_) + self.intercept_
                
                # Calculate gradients
                intercept_derivative = -2 * np.mean(y[idx] - y_hat)
                coef_derivative = -2 * np.dot((y[idx] - y_hat), X[idx])        
                
                # Update parameters
                self.intercept_ -= self.lr * intercept_derivative
                self.coef_ -= self.lr * coef_derivative
                
                # Optionally print the parameters at each epoch (for debugging)
                # if epoch % 10 == 0:  # Print every 10 epochs
                #    print(f"Epoch {epoch}: Coefficients = {self.coef_}, Intercept = {self.intercept_}")
                
    def predict(self, X):
        # Make predictions using the learned parameters
        return np.dot(X, self.coef_) + self.intercept_


In [54]:
bgd = MiniBatchGDRegressor(batch_size=15, learning_rate=0.2, epochs=1000)

In [55]:
bgd.fit(X_train, y_train)  # Train the model first

In [56]:
print("Coefficients after fitting:", bgd.coef_)
print("Intercept after fitting:", bgd.intercept_)

Coefficients after fitting: [ -27.49553122 -179.95048984  547.04682564  329.91487961 -835.61588712
  512.39768309  138.37870182  100.90841896  826.95921069   47.1320404 ]
Intercept after fitting: 155.09519676674572


In [57]:
y_pred_own = bgd.predict(X_test)  # Now make predictions
y_pred_own

array([155.95916567, 212.41113851, 128.31382132, 109.23470433,
       262.45739916, 259.73932118, 121.81747761, 121.95045289,
       105.41458501, 195.83886602, 145.98861924, 173.59329042,
       177.07200274, 135.41103311, 297.20919268,  98.50075006,
       214.65703938, 158.00708481, 138.5204659 , 123.32959206,
       155.81455245, 170.65383892, 150.68989283, 179.29403221,
       135.60078945, 220.72118852, 202.63065624,  98.59291092,
        55.5249171 , 233.3638608 , 242.80015973, 114.57333183,
        71.68037496,  99.18539618, 203.8795514 , 167.59690785,
       163.64169263, 194.09164826, 117.30817711, 229.51669642,
       149.88517796, 125.25172736, 197.78328994, 192.14619507,
       183.06362845, 152.52771142, 166.95030108, 277.74573649,
       103.60846455, 164.45788762, 260.08293246, 144.3365896 ,
       155.84668921, 108.47945506, 197.02024894,  81.95106175,
       119.04915943,  72.72697852, 155.65398787, 167.50191062,
       170.07172804, 160.62694891, 100.89324831, 240.11

In [58]:
r2_score(y_test, y_pred_own)

0.43840275996211453

# Compare with sklearn 

In [60]:
from sklearn.linear_model  import SGDRegressor

In [61]:
sgd = SGDRegressor(learning_rate='constant', eta0=0.1)

In [71]:
# Due to sklearn don't have own parameter 
# we need to create it

batch_size = 50
epochs = 100

for i in range(epochs):
    idx = random.sample(range(X_train.shape[0]), batch_size)
    # This take 1 epochs on one iteration
    sgd.partial_fit(X_train[idx],y_train[idx])

In [72]:
sgd.coef_

array([  20.76572666, -159.61838748,  474.38658285,  286.3146144 ,
        -28.00133166,  -99.21757073, -190.87413876,  102.28630876,
        424.46584783,  110.01970417])

In [73]:
sgd.intercept_

array([159.65791505])

In [74]:
y_pred_sk = sgd.predict(X_test)

In [75]:
r2_score(y_test, y_pred_sk)

0.44811692832342687

In [76]:
# compare our and sklearn model

# Now compare with the model 
# nearly all same

print(f"Linear Regression r^2 Score :{r2_score(y_test, y_pred)} \n")
print(f"Our own Regression r^2 Score :{r2_score(y_test, y_pred_own)} \n")
print(f"SGDRegressor r^2 Score :{r2_score(y_test, y_pred_sk)} \n")

Linear Regression r^2 Score :0.4399338661568968 

Our own Regression r^2 Score :0.43840275996211453 

SGDRegressor r^2 Score :0.44811692832342687 

